File: test/c++11/intel_x64/cr3.cpp
#include <pal/control_register/cr3.h>
#include <stdio.h>

void test_cr3_compile(void)
{
    // Register accessors
    pal::cr3::set(0xA55A5AA5);
    auto value = pal::cr3::get();

    // Field accessors
    pal::cr3::pcd::enable();
    pal::cr3::pcd::enable(value);
    pal::cr3::pcd::disable();
    pal::cr3::pcd::disable(value);
    pal::cr3::pcd::is_enabled();
    pal::cr3::pcd::is_enabled(value);
    pal::cr3::pcd::is_disabled();
    pal::cr3::pcd::is_disabled(value);
    pal::cr3::page_directory_base::get();
    pal::cr3::page_directory_base::get(value);
    pal::cr3::page_directory_base::set(0x0);
    pal::cr3::page_directory_base::set(0x0, value);

    // Printers
    pal::cr3::print(printf);
    pal::cr3::print(printf, value);
    pal::cr3::pcd::print(printf);
    pal::cr3::pcd::print(printf, value);
    pal::cr3::page_directory_base::print(printf);
    pal::cr3::page_directory_base::print(printf, value);
}

File: test/rust/intel/src/cr3.rs
use pal;

pub fn test_cr3_compile()
{
    // Register accessors
    pal::control_register::cr3::set(0xA55A5AA5);
    let mut value = pal::control_register::cr3::get();

    // Field accessors
    pal::control_register::cr3::enable_pcd();
    pal::control_register::cr3::enable_pcd_in_value(&mut value);
    pal::control_register::cr3::disable_pcd();
    pal::control_register::cr3::disable_pcd_in_value(&mut value);
    pal::control_register::cr3::pcd_is_enabled();
    pal::control_register::cr3::pcd_is_enabled_in_value(value);
    pal::control_register::cr3::pcd_is_disabled();
    pal::control_register::cr3::pcd_is_disabled_in_value(value);
    pal::control_register::cr3::get_page_directory_base();
    pal::control_register::cr3::get_page_directory_base_from_value(value);
    pal::control_register::cr3::set_page_directory_base(0x0);
    pal::control_register::cr3::set_page_directory_base_in_value(0x0, &mut value);

    // Printers
    pal::control_register::cr3::print();
    pal::control_register::cr3::print_from_value(value);
    pal::control_register::cr3::print_pcd();
    pal::control_register::cr3::print_pcd_from_value(value);
    pal::control_register::cr3::print_page_directory_base();
    pal::control_register::cr3::print_page_directory_base_from_value(value);
}
